include/deal.II/base/discrete_time.h

```diff
@@ -0,0 +1,190 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2019 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_discrete_time_h
+#define dealii_discrete_time_h
+
+#include <deal.II/base/config.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * Provides a means to keep track of the simulation time in a time-dependent
+ * simulation. It manages stepping forward from an initial time to a final
+ * time. It also allows adjusting the time step size during the simulation.
+ * It is guaranteed that at all times the current simulation time is in the
+ * closed interval between the start time and the end time.
+ *
+ * You can loop over all of the time steps by using a for loop
+ * @code
+ *   for (DiscreteTime time(0., 1., 0.1);
```

Comment:
Maybe make this more obvious (and mention it in a comment):
```suggestion
 *   for (DiscreteTime time(0., 1., 0.3);
```
The issue here is that stepping forward by 0.3 every time doesn't hit the end time, but this class will do the right thing and adjust time step sizes at the end so that we do get to exactly the end time.

Reply:
@jppelteret  Yes. I have some specific language in my head with which to update the documentation of this class once merged -- but I think this has been dragging on for long enough that it's time to get it merged and make other improvements in follow-ups.